File: 2023/day_3/main.go
package main


import (
    "fmt"
    "os"
    "strings"
    "strconv"
)

func isDigit(s string) bool {
    switch s {
    case "0", "1", "2", "3", "4", "5", "6", "7", "8", "9":
        return true
    }
    return false
}

func getEightNeighbors(matrix [][]string, i, j int) []string {
    neighbors := make([]string, 0)
    for k := i-1; k <= i+1; k++ {
        for l := j-1; l <= j+1; l++ {
            if k == i && l == j {
                continue
            }
            if k < 0 || k >= len(matrix) || l < 0 || l >= len(matrix[0]) {
                continue
            }
            neighbors = append(neighbors, matrix[k][l])
        }
    }
    return neighbors
}

func part1() {
    content, err := os.ReadFile("input.txt")
    if err != nil {
        fmt.Println("Error reading file")
        return
    }
    lines := strings.Split(string(content), "\n")
    lines = lines[:len(lines)-1]
    matrix := make([][]string, len(lines))
    for i, line := range lines {
        matrix[i] = strings.Split(line, "")
    }

    listOfNumbers := make([]int, 0)
    number := ""
    symbols := make([]string, 0)
    for i, row := range matrix {
        for j, element := range row {
            if !isDigit(element) {
                if number != "" && len(symbols) > 0 {
                    numberInt, _ := strconv.Atoi(number)
                    listOfNumbers = append(listOfNumbers, numberInt)
                }
                number = ""
                symbols = make([]string, 0)
                continue
            }
            number += element
            neighbors := getEightNeighbors(matrix, i, j)
            symbols = append(symbols, neighbors...)
            for k := 0; k < len(symbols); k++ {
                if symbols[k] == "." || isDigit(symbols[k]) {
                    symbols = append(symbols[:k], symbols[k+1:]...)
                    k--
                }
            }
        }
    }
    if number != "" && len(symbols) > 0 {
        numberInt, _ := strconv.Atoi(number)
        listOfNumbers = append(listOfNumbers, numberInt)
    }
    sum := 0
    for _, number := range listOfNumbers {
        sum += number
    }
    fmt.Println(sum)
}

func getAsteriskNeighbor(matrix [][]string, i, j int) (int, int) {
    for k := i-1; k <= i+1; k++ {
        for l := j-1; l <= j+1; l++ {
            if k == i && l == j {
                continue
            }
            if k < 0 || k >= len(matrix) || l < 0 || l >= len(matrix[0]) {
                continue
            }
            if matrix[k][l] == "*" {
                return k, l
            }
        }
    }
    return -1, -1
}

func part2() {
    content, err := os.ReadFile("input.txt")
    if err != nil {
        fmt.Println("Error reading file")
        return
    }
    lines := strings.Split(string(content), "\n")
    lines = lines[:len(lines)-1]
    matrix := make([][]string, len(lines))
    for i, line := range lines {
        matrix[i] = strings.Split(line, "")
    }

    number := ""
    coordinate := make([]int, 2)
    coordinates := make([][]int, 0)
    numbers := make([]int, 0)
    for i, row := range matrix {
        for j, element := range row {
            if !isDigit(element) {
                if number != "" && coordinate != nil {
                    numberInt, _ := strconv.Atoi(number)
                    numbers = append(numbers, numberInt)
                    coordinates = append(coordinates, coordinate)
                }
                number = ""
                coordinate = make([]int, 2)
                continue
            }
            number += element
            asteriskI, asteriskJ := getAsteriskNeighbor(matrix, i, j)
            if asteriskI != -1 && asteriskJ != -1 {
                coordinate[0] = asteriskI
                coordinate[1] = asteriskJ
            }
        }
    }
    if number != "" && coordinate != nil {
        numberInt, _ := strconv.Atoi(number)
        numbers = append(numbers, numberInt)
        coordinates = append(coordinates, coordinate)
    }
    
    groups := make(map[[2]int][]int)

	// Step 2: Iterate through the lists
	for idx, coord := range coordinates {
		num := numbers[idx]
		groups[[2]int{coord[0], coord[1]}] = append(groups[[2]int{coord[0], coord[1]}], num)
	}

	// Step 3: Extract the lists
	groupedNumbers := make([][]int, 0, len(groups))
	for _, nums := range groups {
		groupedNumbers = append(groupedNumbers, nums)
	}

    sum := 0
    for _, group := range groupedNumbers {
        if len(group) == 2 {
            sum += group[0] * group[1]
        }
    }
    fmt.Println(sum)
}

func main() {
    part1()
    part2()
}

File: 2023/day_3/main.py
def get_eight_neighbors(x: int, y: int, matrix: list[list[str]]) -> list[str]:
    neighbors: list[str] = []
    for i in range(x - 1, x + 2):
        for j in range(y - 1, y + 2):
            try:
                neighbors.append(matrix[i][j])
            except IndexError:
                neighbors.append("")
    return neighbors


def part_1() -> None:
    with open('input.txt', 'r') as f:
        lines: list[str] = f.read().splitlines()

    matrix: list[list[str]] = []
    for line in lines:
        matrix.append(list(line))

    list_of_numbers: list[int] = []
    number: str = ""
    symbols: set[str] = set()
    for i, row in enumerate(matrix):
        for j, element in enumerate(row):
            if not element.isdigit():
                if number and len(symbols):
                    list_of_numbers.append(int(number))
                number = ""
                symbols = set()
                continue

            number += element
            neighbors: list[str] = get_eight_neighbors(i, j, matrix)
            symbols.update(neighbors)
            symbols.difference_update(set([str(i) for i in range(10)] + ['.'] + ['']))

    if number and len(symbols):
        list_of_numbers.append(int(number))

    print(sum(list_of_numbers))


def get_asterisk_neighbor(x: int, y: int, matrix: list[list[str]]) -> tuple[int, int] | None:
    for i in range(x - 1, x + 2):
        for j in range(y - 1, y + 2):
            try:
                if matrix[i][j] == '*':
                    return i, j
            except IndexError:
                pass
    return None


def part_2() -> None:
    with open('input.txt', 'r') as f:
        lines: list[str] = f.read().splitlines()

    matrix: list[list[str]] = []
    for line in lines:
        matrix.append(list(line))

    number: str | None = None
    coordinate: tuple[int, int] | None = None
    coordinates: list[tuple[int, int]] = []
    numbers: list[int] = []
    for i, row in enumerate(matrix):
        for j, element in enumerate(row):
            if not element.isdigit():
                if number is not None and coordinate is not None:
                    coordinates.append(coordinate)
                    numbers.append(int(number))
                number = None
                coordinate = None
                continue

            if number is None:
                number = ""
            number += element
            asterisk_neighbor: tuple[int, int] | None = get_asterisk_neighbor(i, j, matrix)
            if asterisk_neighbor is not None:
                coordinate = asterisk_neighbor

    if number is not None and coordinate is not None:
        numbers.append(int(number))

    groups = {}
    for num, coord in zip(numbers, coordinates):
        if coord not in groups:
            groups[coord] = []
        groups[coord].append(num)

    grouped_numbers = list(groups.values())

    sum: int = 0
    for group in grouped_numbers:
        if len(group) == 2:
            sum += group[0] * group[1]

    print(sum)


if __name__ == '__main__':
    part_1()
    part_2()
